classes/publication/Repository.php

```diff
@@ -148,21 +148,33 @@ public function version(Publication $publication): int
     /** @copydoc \PKP\publication\Repository::setStatusOnPublish() */
     protected function setStatusOnPublish(Publication $publication)
     {
-        // A publication may be scheduled in a future issue. In such cases,
-        // the `datePublished` should remain empty and the status should be set to
-        // scheduled.
-        //
-        // If there is no assigned issue, the journal may be using a continuous
-        // publishing model in which articles are published right away.
+
+        // A publication may be published without an issue. Use current date if no
+        // custom `datePublished` is set. Status is set to published.
+
+        // A publication may be published in a published issue.
+        // In such cases the `datePublished` should be what the user has entered
+        // or fall back to the issue published date. Status should be set to published.
+
+        // A publication may also be scheduled in a future issue. In such cases,
+        // the the status should be set to scheduled and `datePublished`
+        // should remain what it is.
+
         $issue = Repo::issue()->get($publication->getData('issueId'));
-        if ($issue && !$issue->getData('published')) {
-            $publication->setData('datePublished', null);
-            $publication->setData('status', Submission::STATUS_SCHEDULED);
-        } else {
-            $publication->setData('status', Submission::STATUS_PUBLISHED);
+
+        if (!$issue) {
             if (!$publication->getData('datePublished')) {
                 $publication->setData('datePublished', Core::getCurrentDate());
             }
+            $publication->setData('status', Submission::STATUS_PUBLISHED);
+        } elseif ($issue && $issue->getData('published')) {
+            // If no publication datePublished is given, use the issue datePublished
+            if (!$publication->getData('datePublished')) {
+                $publication->setData('datePublished', $issue->getData('datePublished'));
+            }
+            $publication->setData('status', Submission::STATUS_PUBLISHED);
+        } else {
+            $publication->setData('status', Submission::STATUS_SCHEDULED);
```

Comment:
This looks good. However, I'd like to restructure the conditional block a little. Right now the `else` block corresponds to a very specific case: a future issue. But it's not that clear from reading the code. Can we structure this to be a little more explicit for reading?

```php
if ($issue) {
    if ($issue->getData('published')) {
        $publication->setData('status', Submission::STATUS_PUBLISHED);
        if (!$publication->getData('datePublished')) {
            $publication->setData('datePublished', $issue->getData('datePublished'));
        }
    } else {
        $publication->setData('status', Submission::STATUS_SCHEDULED);
    }
} else {
    $publication->setData('status', Submission::STATUS_PUBLISHED);
    if (!$publication->getData('datePublished')) {
        $publication->setData('datePublished', Core::getCurrentDate());
    }
}
```

It results in a little more conditional block nesting. But I think it's easier to interpret what is happening in each block.

Reply:
Looks good! Just want to note that again the last `else` is something that the current code base will never trigger. You will not get here without having an `issue` set. But it is definitely something we have been working towards. Just thinking that do we want to have this "sleeping code" here already now?

Reply:
That's a fair point, but if we left the `else` block out, we'd have an "invisible" condition. Calling `setStatusOnPublish()` for such a publication would do nothing. I'd prefer to fall back to a sensible default behavior that does what the function says it does (changing the status when something is published).

Reply:
sure agree!